keps/sig-node/20190717-seccomp-ga.md

```diff
@@ -0,0 +1,361 @@
+
+```
+seccomp.security.alpha.kubernetes.io/pod: {unconfined,docker/default,localhost/<path>}
+container.seccomp.security.alpha.kubernetes.io/<container_name>: ...
+```
+
+The feature has been more or less unchanged ever since. Also note that the addition predates feature
+gates or our modern concept of feature lifecycle. So, even though the annotations include `alpha` in
+the key, this is entirely useable on any production GA cluster.
+
+There have been multiple attempts to [change the default
+profile](https://github.com/kubernetes/kubernetes/issues/39845) or [formally spec the Kubernetes
+seccomp profile](https://github.com/kubernetes/kubernetes/issues/39128), but both efforts were
+abandoned due to friction and lack of investment.
+
+Despite the `alpha` label, I think this feature needs to be treated as GA, and we're doing our users
+a disservice by leaving it in this weird limbo state. As much as I would like to see seccomp support
+fully fleshed out, if we block GA on those enhancements we will remain stuck in the current state
+indefinitely. Therefore, I'm proposing we do the absolute minimum to clean up the current
+implementation all accurately declare the feature "GA". Future enhancements can follow the standard
+alpha -> beta -> GA feature process.
+
+_NOTE: AppArmor is in a very similar state, but with some subtle differences. Promoting AppArmor to
+GA will be covered by a separate KEP._
+
+### Goals
+
+- Declare seccomp GA
+- Fully document and formally spec the feature support
+- Migrate the annotations to standard API fields
+- Deprecate the seccomp annotations
+
+### Non-Goals
+
+This KEP proposes the absolute minimum to get seccomp to GA, therefore all functional enhancements
+are out of scope, including:
+
+- Changing the default seccomp profile from `unconfined`
+- Defining any standard "Kubernetes branded" seccomp profiles
+- Formally speccing the seccomp profile format in Kubernetes
+- Providing mechanisms for loading profiles from outside the static seccomp node directory
+- Changing the semantics around seccomp support
+
+## Proposal
+
+### API
+
+The seccomp API will be functionally equivalent to the current alpha API. This includes the Pod API,
+which specifies what profile the pod & containers run with, and the PodSecurityPolicy API which
+specifies allowed profiles & a default profile.
+
+#### Pod API
+
+The Pod Seccomp API is immutable.
+
+```go
+type PodSecurityContext struct {
+    ...
+    // The seccomp options to use by the containers in this pod.
+    // +optional
+    Seccomp  *SeccompOptions
+    ...
+}
+
+type SecurityContext struct {
+    ...
+    // The seccomp options to use by this container. If seccomp options are
+    // provided at both the pod & container level, the container options
+    // override the pod options.
+    // +optional
+    Seccomp  *SeccompOptions
+    ...
+}
+
+type SeccompOptions struct {
+    // The seccomp profile to run with.
+    SeccompProfile
```

Comment:
was this intended to be embedded rather than something like `Profile SeccompProfile`? what other options do you anticipate?

Reply:
Yes, I modeled it off the `Volume` type.

The only option that comes to mind is the fallback strategy. Currently, if the seccomp options can't be enforced the node silently ignores them. In contrast, when AppArmor can't be enforced the pod is held in a `Blocked` state.

From a security perspective, silently failing could leave a potential security hole. From a portability perspective, it's nice to be able to choose the best practices options with a best-effort approach. Hence it might make sense to give the user the choice.

Reply:
what happens for nodes/runtimes that don't enable seccomp is worth documenting, at least

Reply:
maybe just `securityContext.seccompProfile` instead of embedding/nesting (can prefix other seccomp fields with `seccomp` if they are needed in the future